8342095: Add autovectorizer support for subword vector casts

jaskarth opened the pull request:
Hi all,
This patch adds initial support for the autovectorizer to generate conversions between subword types. Currently, when superword sees two packs that have different basic types, it discards them and bails out of vectorization. This patch changes the behavior to ask the backend if a cast between the conflicting types is supported, and keeps the pack if it is. Later, when the `VTransform` graph is built, a synthetic cast is emitted when packs requiring casts are detected. Currently, only narrowing casts are supported as I wanted to re-use existing `VectorCastX2Y` logic for the initial version, but adding more conversions is simple and can be done with a subsequent RFE. I have attached a JMH benchmark and got these results on my Zen 3 machine:

```c
                                                  Baseline                    Patch
Benchmark                  (SIZE)  Mode  Cnt    Score    Error  Units   Score    Error  Units    Improvement
VectorSubword.intToByte      1024  avgt   12  200.049 ± 19.787  ns/op   56.228 ± 3.535  ns/op  (3.56x)
VectorSubword.intToShort     1024  avgt   12  179.826 ±  1.539  ns/op   43.332 ± 1.166  ns/op  (4.15x)
VectorSubword.shortToByte    1024  avgt   12  245.580 ±  6.150  ns/op   29.757 ± 1.055  ns/op  (8.25x)
```

I've also added some IR tests and they pass on my linux x64 machine. Thoughts and reviews would be appreciated!

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8342095](https://bugs.openjdk.org/browse/JDK-8342095): Add autovectorizer support for subword vector casts (**Enhancement** - P4)


### Reviewers
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/23413/head:pull/23413` \
`$ git checkout pull/23413`

Update a local copy of the PR: \
`$ git checkout pull/23413` \
`$ git pull https://git.openjdk.org/jdk.git pull/23413/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 23413`

View PR using the GUI difftool: \
`$ git pr show -t 23413`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/23413.diff">https://git.openjdk.org/jdk/pull/23413.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/23413#issuecomment-2629906054)
</details>

jaskarth commented on test/hotspot/jtreg/compiler/vectorization/runner/ArrayTypeConvertTest.java:
```diff
@@ -24,6 +24,7 @@
 
 /*
  * @test
+ * @bug 8183390 8340010 8342095
```
Review note: This test was missing a `@bug` annotation, so I went through the history and added the ones that substantially modified the test.

eme64 commented on src/hotspot/share/opto/vtransform.hpp:
```diff
@@ -522,6 +522,22 @@ class VTransformStoreVectorNode : public VTransformMemVectorNode {
+class VTransformCastNode : public VTransformNode {
+private:
+  uint _vlen;
+  BasicType _from_bt;
+  BasicType _to_bt;
```
Nit: I would consistently use `def` and `use`.

eme64 commented on src/hotspot/share/opto/vtransform.hpp:
```diff
@@ -522,6 +522,22 @@ class VTransformStoreVectorNode : public VTransformMemVectorNode {
   NOT_PRODUCT(virtual const char* name() const override { return "StoreVector"; };)
 };
 
+class VTransformCastNode : public VTransformNode {
```
I think it would be good to make it a `VTransformCastVectorNode`.

eme64 commented on test/hotspot/jtreg/compiler/loopopts/superword/TestSubwordVectorization.java:
```diff
@@ -0,0 +1,155 @@
+    static Object[] setupIntArray() {
+        int[] res = new int[SIZE];
+
+        for (int i = 0; i < SIZE; i++) {
+            res[i] = RANDOM.nextInt();
+        }
```
Can you please use `Generators.java`? It would be great if we can use that in the future, to create more "interesting" input data ;)

eme64 commented on test/hotspot/jtreg/compiler/loopopts/superword/TestSubwordVectorization.java:
```diff
@@ -0,0 +1,155 @@
+    @Test
+    @IR(applyIfCPUFeature = { "avx", "true" },
+        applyIfOr = {"AlignVector", "false", "UseCompactObjectHeaders", "false"},
+        counts = { IRNode.VECTOR_CAST_I2B, IRNode.VECTOR_SIZE_ANY, ">0" })
```
Hmm. It would be great if we could assert the length of the vector. Otherwise we may get less than we could.

Something like this may work:
`IRNode.VECTOR_SIZE + "min(max_int, max_byte)"`

Have you tried that yet?

jaskarth replied:
I think I wrote this test before realizing you can do size checks like in `ArrayTypeConvertTest.java`, I'll make sure to change it to the same system there.

eme64 commented on test/hotspot/jtreg/compiler/vectorization/runner/ArrayTypeConvertTest.java:
```diff
@@ -148,10 +149,9 @@ public int[] signExtensionFromByte() {
-    // the test when this limitation is addressed in the future.
+    @IR(applyIfCPUFeature = { "avx", "true" },
+        applyIfOr = {"AlignVector", "false", "UseCompactObjectHeaders", "false"},
+        counts = { IRNode.VECTOR_CAST_I2S, IRNode.VECTOR_SIZE + "min(max_int, max_short)", ">0" })
```
Ah see, here we have the vector size asserted, good!

eme64 commented on test/micro/org/openjdk/bench/vm/compiler/VectorSubword.java:
```diff
@@ -0,0 +1,73 @@
+        ints = new int[SIZE];
+    }
+
+    @Benchmark
+    public void intToShort() {
+        for (int i = 0; i < SIZE; i++) {
+            shorts[i] = (short) ints[i];
+        }
+    }
+
+    @Benchmark
+    public void intToByte() {
+        for (int i = 0; i < SIZE; i++) {
+            bytes[i] = (byte) ints[i];
+        }
+    }
+
+    @Benchmark
+    public void shortToByte() {
+        for (int i = 0; i < SIZE; i++) {
+            bytes[i] = (byte) shorts[i];
+        }
+    }
+}
```
Ah, these are all casting to smaller types. What about casting to larger types?

jaskarth replied:
Originally I thought that larger type conversions were only available on AVX-512 based on the ad-file, but reading it more carefully now I see that they are indeed supported for AVX! I'll go ahead and generalize the changes further to make casting to larger types supported as well. I think this leaves `char` as the only type not supported, which I can look at in a follow-up RFE.

eme64 replied:
Well it's ok to leave the patch as is, and just have a follow-up RFE that improves things from here. It could make things easier to review.

eme64 replied:
@jaskarth

jaskarth replied:
@eme64 I took a look last night and it looks like supporting casting to larger types (other than `char`) just involves changing the filtering logic in `matcher_x86.hpp` and adding more unit tests, without modifying any of the core logic. I think it could make sense to keep it in the same patch, what do you think? Either way would work with me.

eme64 replied:
I leave it up to you. If it's not much then pack it in the same, else splitting is fine with me too ;)

eme64 commented on test/hotspot/jtreg/compiler/vectorization/runner/ArrayTypeConvertTest.java:
```diff
@@ -161,10 +161,9 @@ public short[] narrowToSigned() {
 
     @Test
-    @IR(failOn = {IRNode.STORE_VECTOR})
-    // Subword vector casts do not work currently, see JDK-8342095.
```
There seem to be other cases in this file that mention `JDK-8342095`. We should probably file a new RFE for those, right?

eme64 reviewed (commented):
Great work, looks generally amazing :blush: 

I left a few comments below.

jaskarth commented:
@eme64 I've pushed a new version that implements widening, as well as modifies the unit tests to use the new `Generators` framework. I've also filed [JDK-8349562](https://bugs.openjdk.org/browse/JDK-8349562) as followup work for `char` types. Let me know what you think!

jaskarth commented:
I also updated the benchmark, and got these results:
```c
                                                  Baseline                    Patch
Benchmark                  (SIZE)  Mode  Cnt    Score    Error  Units  Score    Error  Units  Improvement
VectorSubword.byteToInt      1024  avgt   12  185.700 ± 0.798  ns/op   37.427 ± 0.276  ns/op  (4.96x)
VectorSubword.byteToShort    1024  avgt   12  240.737 ± 1.087  ns/op   23.094 ± 0.502  ns/op  (10.42x)
VectorSubword.intToByte      1024  avgt   12  181.680 ± 0.553  ns/op   49.873 ± 1.613  ns/op  (3.64x)
VectorSubword.intToShort     1024  avgt   12  176.256 ± 1.414  ns/op   43.933 ± 4.310  ns/op  (4.01x)
VectorSubword.shortToByte    1024  avgt   12  245.600 ± 6.217  ns/op   28.426 ± 0.649  ns/op  (8.64x)
VectorSubword.shortToInt     1024  avgt   12  178.364 ± 2.921  ns/op   34.140 ± 0.229  ns/op  (5.22x)
```

eme64 commented on src/hotspot/cpu/x86/matcher_x86.hpp:
```diff
@@ -261,6 +261,29 @@
+      return false;
+    }
+
+    switch (from_bt) {
+      case T_INT:
+      case T_SHORT:
+      case T_BYTE: {
+        return to_bt == T_INT || to_bt == T_SHORT || to_bt == T_BYTE;
+      }
+      default: {
+        return false;
+      }
+    }
```
You could leave a quick comment about why `CHAR` is not yet covered here.

eme64 commented on src/hotspot/share/opto/vtransform.hpp:
```diff
@@ -522,6 +522,22 @@ class VTransformStoreVectorNode : public VTransformMemVectorNode {
+    _vlen(vlen), _from_bt(from_bt), _to_bt(to_bt) {}
+  virtual VTransformApplyResult apply(const VLoopAnalyzer& vloop_analyzer,
+                                      const GrowableArray<Node*>& vnode_idx_to_transformed_node) const override;
+  NOT_PRODUCT(virtual const char* name() const override { return "Cast"; };)
```
```suggestion
  NOT_PRODUCT(virtual const char* name() const override { return "CastVector"; };)
```

eme64 replied:
I know the node is not inheriting from `VTransformVectorNode`, but we can make that happen with a future refactoring I'm already working on.

eme64 commented on test/hotspot/jtreg/compiler/loopopts/superword/TestCompatibleUseDefTypeSize.java:
```diff
@@ -327,12 +327,12 @@ static Object[] test0(byte[] src, char[] dst) {
         applyIfPlatform = {"64-bit", "true"},
-        applyIfCPUFeatureOr = {"sse4.1", "true", "asimd", "true"})
+        applyIf = {"AlignVector", "false"},
+        applyIfCPUFeature = {"avx", "true"})
```
This may be a little nit-picky. But why have a new test-file when this test here was already trying to cover the conversion cases? I think I wrote it back then, and was just too lazy to write all conversion cases. I'd suggest you move your cases up here ;)

eme64 replied:
I think I added these tests when I was reworking `SuperWord::is_velt_basic_type_compatible_use_def`, which you are now touching as well :blush:

jaskarth replied:
I think I didn't realize this test was there when I was working on my patch, but I agree it makes sense to move it there for clarity.

eme64 reviewed (commented):
@jaskarth That looks great! Thanks for the updates!

A little control question:
We now add additional vector operations into the loop body. How do we know this will not lead to a regression in some cases? I think it should not... right? Are the casts no-ops, or could they have some cost to them? Could there be any case where the wins from vectorization would not make up for the extra vector cast node?

jaskarth commented on src/hotspot/cpu/x86/matcher_x86.hpp:
```diff
@@ -261,6 +261,29 @@
+      return false;
+    }
+
+    // Cannot cast to own type
+    if (to_bt == from_bt) {
+      return false;
+    }
```
This is a good point, I think it shouldn't be possible to run into the case where `to_bt` and `from_bt` are the same here so a defensive assert would be better here.

eme64 commented:
@jaskarth just ping me whenever I should have a look again!

jaskarth commented:
@eme64 I think it should be good for another look over! I've addressed your review comments in the last commit.

About the potential for performance degradation, I think it would be unlikely since the code generated by the cast is quite small (as it only needs to truncate or sign-extend) and the patch increases the amount of possible code that can auto-vectorize. The one case that I can think of is that it might cause code that would be otherwise unprofitable to become vectorizable, but that would be because we don't have a cost model yet.

eme64 commented on src/hotspot/cpu/aarch64/matcher_aarch64.hpp:
```diff
@@ -198,6 +198,10 @@
 
+  static bool is_vector_cast_supported(BasicType from_bt, BasicType to_bt) {
+    return false;
+  }
```
Do these simply not exist, or do you just want to leave that to a future RFE so someone else can take care of this?

eme64 commented on src/hotspot/cpu/x86/matcher_x86.hpp:
```diff
@@ -261,6 +261,27 @@
     }
   }
 
+  static bool is_vector_cast_supported(BasicType from_bt, BasicType to_bt) {
```
Why does this not live next to `Matcher::match_rule_supported_vector`, would that not be a better fit?

eme64 commented on src/hotspot/share/opto/superwordVTransformBuilder.cpp:
```diff
@@ -186,6 +186,19 @@ VTransformNode* SuperWordVTransformBuilder::get_or_make_vtnode_vector_input_at_i
+    BasicType use_bt = _vloop_analyzer.types().velt_basic_type(p0);
+
+    // If the use and def types are different, emit a cast node
+    if (use_bt != def_bt && !p0->is_Convert() && Matcher::is_vector_cast_supported(def_bt, use_bt)) {
```
The usual way we check if a vector instruction is implemented is to use `VectorNode::implemented`. Ah, actually there is a `VectorCastNode::implemented`. Why are you not using that one?

jaskarth replied:
This is a good point! I've updated the patch to use `VectorCastNode::implemented` instead. I think I didn't see that function originally.

eme64 reviewed (commented):
I had another quick look, and have some additional questions :)

eme64 commented:
@jaskarth Let me know if there is anything we can help you with here :)

eme64 commented on test/micro/org/openjdk/bench/vm/compiler/VectorSubword.java:
```diff
@@ -0,0 +1,94 @@
+    @Param({"1024"})
+    public int SIZE;
+
+    private byte[] bytes;
+    private short[] shorts;
+    private int[] ints;
```
It would be nice if you covered also `char` and `long`, for completeness :)

eme64 reviewed (changes requested):
This looks really good, thanks @jaskarth for the work you put in!

I have a few more comments below.

jaskarth commented:
@eme64 Thank you for the comments! I've updated the test and benchmark to be more exhaustive, and applied the suggested changes. For the benchmark, I got these results on my machine:
```c
                                                  Baseline                    Patch
Benchmark                  (SIZE)  Mode  Cnt    Score    Error  Units    Score   Error  Units  Improvement
VectorSubword.byteToChar     1024  avgt   12  252.954 ±  4.129  ns/op   24.219 ± 0.453  ns/op  (10.4x)
VectorSubword.byteToInt      1024  avgt   12  194.707 ±  3.584  ns/op   38.353 ± 0.637  ns/op  (5.07x)
VectorSubword.byteToLong     1024  avgt   12   73.645 ±  1.418  ns/op   70.521 ± 0.470  ns/op  (no change)
VectorSubword.byteToShort    1024  avgt   12  252.647 ±  3.738  ns/op   22.664 ± 0.449  ns/op  (11.1x)
VectorSubword.charToByte     1024  avgt   12  236.396 ±  3.893  ns/op  228.710 ± 1.967  ns/op  (no change)
VectorSubword.charToInt      1024  avgt   12  179.673 ±  2.811  ns/op  173.764 ± 1.150  ns/op  (no change)
VectorSubword.charToLong     1024  avgt   12  184.867 ±  3.079  ns/op  177.999 ± 1.312  ns/op  (no change)
VectorSubword.charToShort    1024  avgt   12   24.385 ±  1.822  ns/op   22.375 ± 1.980  ns/op  (no change)
VectorSubword.intToByte      1024  avgt   12  190.949 ±  1.475  ns/op   49.376 ± 1.383  ns/op  (3.86x)
VectorSubword.intToChar      1024  avgt   12  182.862 ±  3.708  ns/op   44.344 ± 4.513  ns/op  (4.12x)
VectorSubword.intToLong      1024  avgt   12   76.072 ±  1.153  ns/op   73.382 ± 0.294  ns/op  (no change)
VectorSubword.intToShort     1024  avgt   12  184.362 ±  1.938  ns/op   45.556 ± 3.323  ns/op  (4.04x)
VectorSubword.longToByte     1024  avgt   12  150.766 ±  3.475  ns/op  146.651 ± 0.742  ns/op  (no change)
VectorSubword.longToChar     1024  avgt   12  121.764 ±  1.323  ns/op  117.068 ± 1.891  ns/op  (no change)
VectorSubword.longToInt      1024  avgt   12   83.761 ±  2.140  ns/op   82.084 ± 0.930  ns/op  (no change)
VectorSubword.longToShort    1024  avgt   12  132.293 ± 23.046  ns/op  115.883 ± 0.834  ns/op  (+ 12.4%)
VectorSubword.shortToByte    1024  avgt   12  253.387 ±  5.972  ns/op   27.591 ± 1.311  ns/op  (9.18x)
VectorSubword.shortToChar    1024  avgt   12   21.446 ±  1.914  ns/op   20.608 ± 1.593  ns/op  (no change)
VectorSubword.shortToInt     1024  avgt   12  187.109 ±  3.372  ns/op   36.818 ± 0.989  ns/op  (5.08x)
VectorSubword.shortToLong    1024  avgt   12   75.448 ±  0.930  ns/op   72.835 ± 0.507  ns/op  (no change)
```
Interestingly, even though the `longToType` methods are now vectorizable, the performance difference is very small. I think it could be because on my AVX2 machine, it can only process 4 elements per iteration and the overhead of the conversion is fairly high. It's interesting that the `longToInt` method is faster than the rest, I'm curious if something can be done on the backend to improve the speed. I think there could also be potential speed improvements on platforms with wider vectors, like AVX512.

eme64 commented on test/hotspot/jtreg/compiler/vectorization/runner/ArrayShiftOpTest.java:
```diff
@@ -1,6 +1,7 @@
 /*
  * Copyright (c) 2022, 2023, Arm Limited. All rights reserved.
  * Copyright (c) 2024, 2025, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2025, Rivos Inc. All rights reserved.
```
Looks lie a random copyright update. Did you mean to do that?

eme64 commented on src/hotspot/share/opto/superword.cpp:
```diff
@@ -2352,8 +2352,18 @@ bool SuperWord::is_velt_basic_type_compatible_use_def(Node* use, Node* def) cons
+  }
+
+  // Subword cast: Element sizes differ, but the platform supports a cast to change the def shape to the use shape.
+
```
```suggestion
```

eme64 commented:
@jaskarth I'm running some internal testing now, feel free to ping me after the weekend for results!

eme64 commented:
I ran the benchmarks on my `avx512` laptop. I noticed it would take 16 min, so I cut it short like this, so it only takes a little more than 5 min:
```
   35 @Warmup(iterations = 2, time = 1, timeUnit = TimeUnit.SECONDS)
   36 @Measurement(iterations = 3, time = 1, timeUnit = TimeUnit.SECONDS)
```

![image](https://github.com/user-attachments/assets/2ad2a043-55eb-4970-b597-9c30425d18be)

@jaskarth I think your "improvement" in `VectorSubword.longToShort` is just due to noise on your master run, the improvement lies within the error.

eme64 commented:
@jaskarth I ran it with `perf`:

`make test TEST="micro:VectorSubword.charToByte" CONF=linux-x64 TEST_VM_OPTS="-XX:-UseSuperWord" MICRO="OPTIONS=-prof perfasm"`

No SuperWord, about 90%+ of the time are spent in the main loop, 8x unrolled:
```
   1.13%  ↗│     0x00007fc0b01d79b0:   movslq %r11d,%r14                   ;*bastore {reexecute=0 rethrow=0 return_oop=0}
          ││                                                               ; - org.openjdk.bench.vm.compiler.VectorSubword::charToByte@22 (line 76)
          ││                                                               ; - org.openjdk.bench.vm.compiler.jmh_generated.VectorSubword_charToByte_jmhTest::charToByte_avgt_jmhStub@15 (line 190)
   2.44%  ││     0x00007fc0b01d79b3:   movzwl 0x10(%rdi,%r14,2),%r10d      ;*caload {reexecute=0 rethrow=0 return_oop=0}
          ││                                                               ; - org.openjdk.bench.vm.compiler.VectorSubword::charToByte@20 (line 76)
          ││                                                               ; - org.openjdk.bench.vm.compiler.jmh_generated.VectorSubword_charToByte_jmhTest::charToByte_avgt_jmhStub@15 (line 190)
   2.73%  ││     0x00007fc0b01d79b9:   mov    %r10b,0x10(%r9,%r14,1)       ;*bastore {reexecute=0 rethrow=0 return_oop=0}
          ││                                                               ; - org.openjdk.bench.vm.compiler.VectorSubword::charToByte@22 (line 76)
          ││                                                               ; - org.openjdk.bench.vm.compiler.jmh_generated.VectorSubword_charToByte_jmhTest::charToByte_avgt_jmhStub@15 (line 190)
   5.89%  ││     0x00007fc0b01d79be:   movzwl 0x1e(%rdi,%r14,2),%r10d
   2.35%  ││     0x00007fc0b01d79c4:   movzwl 0x1c(%rdi,%r14,2),%esi
   1.51%  ││     0x00007fc0b01d79ca:   movzwl 0x1a(%rdi,%r14,2),%r8d
   5.34%  ││     0x00007fc0b01d79d0:   movzwl 0x18(%rdi,%r14,2),%edx
   1.16%  ││     0x00007fc0b01d79d6:   movzwl 0x16(%rdi,%r14,2),%ebx
   1.77%  ││     0x00007fc0b01d79dc:   movzwl 0x14(%rdi,%r14,2),%ebp
   2.51%  ││     0x00007fc0b01d79e2:   movzwl 0x12(%rdi,%r14,2),%eax       ;*caload {reexecute=0 rethrow=0 return_oop=0}
          ││                                                               ; - org.openjdk.bench.vm.compiler.VectorSubword::charToByte@20 (line 76)
          ││                                                               ; - org.openjdk.bench.vm.compiler.jmh_generated.VectorSubword_charToByte_jmhTest::charToByte_avgt_jmhStub@15 (line 190)
   5.73%  ││     0x00007fc0b01d79e8:   mov    %al,0x11(%r9,%r14,1)
  13.70%  ││     0x00007fc0b01d79ed:   mov    %bpl,0x12(%r9,%r14,1)
   6.72%  ││     0x00007fc0b01d79f2:   mov    %bl,0x13(%r9,%r14,1)
   9.84%  ││     0x00007fc0b01d79f7:   mov    %dl,0x14(%r9,%r14,1)
   6.01%  ││     0x00007fc0b01d79fc:   mov    %r8b,0x15(%r9,%r14,1)
   6.05%  ││     0x00007fc0b01d7a01:   mov    %sil,0x16(%r9,%r14,1)
   5.24%  ││     0x00007fc0b01d7a06:   mov    %r10b,0x17(%r9,%r14,1)       ;*bastore {reexecute=0 rethrow=0 return_oop=0}
          ││                                                               ; - org.openjdk.bench.vm.compiler.VectorSubword::charToByte@22 (line 76)
          ││                                                               ; - org.openjdk.bench.vm.compiler.jmh_generated.VectorSubword_charToByte_jmhTest::charToByte_avgt_jmhStub@15 (line 190)
          ││                                                               ;   {other}
  11.16%  ││     0x00007fc0b01d7a0b:   add    $0x8,%r11d                   ;*iinc {reexecute=0 rethrow=0 return_oop=0}
          ││                                                               ; - org.openjdk.bench.vm.compiler.VectorSubword::charToByte@23 (line 75)
          ││                                                               ; - org.openjdk.bench.vm.compiler.jmh_generated.VectorSubword_charToByte_jmhTest::charToByte_avgt_jmhStub@15 (line 190)
   0.23%  ││     0x00007fc0b01d7a0f:   cmp    %ecx,%r11d
          ╰│     0x00007fc0b01d7a12:   jl     0x00007fc0b01d79b0           ;*goto {reexecute=0 rethrow=0 return_oop=0}
           │                                                               ; - org.openjdk.bench.vm.compiler.VectorSubword::charToByte@26 (line 75)
           │                                                               ; - org.openjdk.bench.vm.compiler.jmh_generated.VectorSubword_charToByte_jmhTest::charToByte_avgt_jmhStub@15 (line 190)
```

With SuperWord this happens:
```
   2.91%  ↗  0x00007f5f801d6d71:   mov    %r11d,%r8d                   ;*aload_0 {reexecute=0 rethrow=0 return_oop=0}
          │                                                            ; - org.openjdk.bench.vm.compiler.VectorSubword::charToByte@10 (line 76)
          │                                                            ; - org.openjdk.bench.vm.compiler.jmh_generated.VectorSubword_charToByte_jmhTest::charToByte_avgt_jmhStub@15 (line 190)
   0.78%  │  0x00007f5f801d6d74:   vmovd  %r8d,%xmm4
   0.81%  │  0x00007f5f801d6d79:   movslq %r8d,%r14                    ;*bastore {reexecute=0 rethrow=0 return_oop=0}
          │                                                            ; - org.openjdk.bench.vm.compiler.VectorSubword::charToByte@22 (line 76)
          │                                                            ; - org.openjdk.bench.vm.compiler.jmh_generated.VectorSubword_charToByte_jmhTest::charToByte_avgt_jmhStub@15 (line 190)
          │  0x00007f5f801d6d7c:   movzwl 0x10(%r13,%r14,2),%r10d      ;*caload {reexecute=0 rethrow=0 return_oop=0}
          │                                                            ; - org.openjdk.bench.vm.compiler.VectorSubword::charToByte@20 (line 76)
          │                                                            ; - org.openjdk.bench.vm.compiler.jmh_generated.VectorSubword_charToByte_jmhTest::charToByte_avgt_jmhStub@15 (line 190)
   9.42%  │  0x00007f5f801d6d82:   mov    %r10b,0x10(%rbp,%r14,1)      ;*bastore {reexecute=0 rethrow=0 return_oop=0}
          │                                                            ; - org.openjdk.bench.vm.compiler.VectorSubword::charToByte@22 (line 76)
          │                                                            ; - org.openjdk.bench.vm.compiler.jmh_generated.VectorSubword_charToByte_jmhTest::charToByte_avgt_jmhStub@15 (line 190)
   4.40%  │  0x00007f5f801d6d87:   movzwl 0x4e(%r13,%r14,2),%r10d
   1.10%  │  0x00007f5f801d6d8d:   vmovd  %r10d,%xmm3
   0.16%  │  0x00007f5f801d6d92:   movzwl 0x4c(%r13,%r14,2),%r10d
   0.29%  │  0x00007f5f801d6d98:   vmovd  %r10d,%xmm6
   2.14%  │  0x00007f5f801d6d9d:   movzwl 0x4a(%r13,%r14,2),%r10d
          │  0x00007f5f801d6da3:   vmovd  %r10d,%xmm5
   2.17%  │  0x00007f5f801d6da8:   movzwl 0x48(%r13,%r14,2),%r10d
   0.03%  │  0x00007f5f801d6dae:   vmovd  %r10d,%xmm8
   1.91%  │  0x00007f5f801d6db3:   movzwl 0x46(%r13,%r14,2),%r10d
          │  0x00007f5f801d6db9:   vmovd  %r10d,%xmm7
   3.76%  │  0x00007f5f801d6dbe:   movzwl 0x44(%r13,%r14,2),%r10d      ;   {other}
   0.19%  │  0x00007f5f801d6dc4:   vmovd  %r10d,%xmm10
   1.07%  │  0x00007f5f801d6dc9:   movzwl 0x42(%r13,%r14,2),%r10d
          │  0x00007f5f801d6dcf:   vmovd  %r10d,%xmm9
   1.68%  │  0x00007f5f801d6dd4:   movzwl 0x40(%r13,%r14,2),%r10d
          │  0x00007f5f801d6dda:   vmovd  %r10d,%xmm12
   1.81%  │  0x00007f5f801d6ddf:   movzwl 0x3e(%r13,%r14,2),%r10d
          │  0x00007f5f801d6de5:   vmovd  %r10d,%xmm11
   2.17%  │  0x00007f5f801d6dea:   movzwl 0x3c(%r13,%r14,2),%r10d
   0.16%  │  0x00007f5f801d6df0:   vmovd  %r10d,%xmm14
   3.37%  │  0x00007f5f801d6df5:   movzwl 0x3a(%r13,%r14,2),%r10d
          │  0x00007f5f801d6dfb:   vmovd  %r10d,%xmm13
   1.98%  │  0x00007f5f801d6e00:   movzwl 0x38(%r13,%r14,2),%r10d
          │  0x00007f5f801d6e06:   vmovd  %r10d,%xmm16
   1.59%  │  0x00007f5f801d6e0c:   movzwl 0x36(%r13,%r14,2),%r10d
          │  0x00007f5f801d6e12:   vmovd  %r10d,%xmm15
   2.49%  │  0x00007f5f801d6e17:   movzwl 0x34(%r13,%r14,2),%r10d
   0.13%  │  0x00007f5f801d6e1d:   vmovd  %r10d,%xmm18
   1.94%  │  0x00007f5f801d6e23:   movzwl 0x32(%r13,%r14,2),%r10d
          │  0x00007f5f801d6e29:   vmovd  %r10d,%xmm17
   2.43%  │  0x00007f5f801d6e2f:   movzwl 0x30(%r13,%r14,2),%r10d
          │  0x00007f5f801d6e35:   vmovd  %r10d,%xmm20
   1.62%  │  0x00007f5f801d6e3b:   movzwl 0x2e(%r13,%r14,2),%r10d
   0.84%  │  0x00007f5f801d6e41:   vmovd  %r10d,%xmm19
   2.36%  │  0x00007f5f801d6e47:   movzwl 0x2c(%r13,%r14,2),%r10d
   0.06%  │  0x00007f5f801d6e4d:   vmovd  %r10d,%xmm22
   3.17%  │  0x00007f5f801d6e53:   movzwl 0x2a(%r13,%r14,2),%r10d
          │  0x00007f5f801d6e59:   vmovd  %r10d,%xmm21
   2.04%  │  0x00007f5f801d6e5f:   movzwl 0x28(%r13,%r14,2),%r10d
          │  0x00007f5f801d6e65:   vmovd  %r10d,%xmm24
   1.42%  │  0x00007f5f801d6e6b:   movzwl 0x26(%r13,%r14,2),%r10d
          │  0x00007f5f801d6e71:   vmovd  %r10d,%xmm23
   1.72%  │  0x00007f5f801d6e77:   movzwl 0x24(%r13,%r14,2),%esi
   0.03%  │  0x00007f5f801d6e7d:   movzwl 0x22(%r13,%r14,2),%r10d
   0.03%  │  0x00007f5f801d6e83:   movzwl 0x20(%r13,%r14,2),%r11d
          │  0x00007f5f801d6e89:   movzwl 0x1e(%r13,%r14,2),%r9d
          │  0x00007f5f801d6e8f:   movzwl 0x1c(%r13,%r14,2),%r8d
   0.03%  │  0x00007f5f801d6e95:   movzwl 0x1a(%r13,%r14,2),%ebx
   0.39%  │  0x00007f5f801d6e9b:   movzwl 0x18(%r13,%r14,2),%ecx
          │  0x00007f5f801d6ea1:   movzwl 0x16(%r13,%r14,2),%edx
   1.85%  │  0x00007f5f801d6ea7:   movzwl 0x14(%r13,%r14,2),%edi
          │  0x00007f5f801d6ead:   movzwl 0x12(%r13,%r14,2),%eax       ;*caload {reexecute=0 rethrow=0 return_oop=0}
          │                                                            ; - org.openjdk.bench.vm.compiler.VectorSubword::charToByte@20 (line 76)
          │                                                            ; - org.openjdk.bench.vm.compiler.jmh_generated.VectorSubword_charToByte_jmhTest::charToByte_avgt_jmhStub@15 (line 190)
   0.06%  │  0x00007f5f801d6eb3:   mov    %al,0x11(%rbp,%r14,1)
   0.03%  │  0x00007f5f801d6eb8:   mov    %dil,0x12(%rbp,%r14,1)
          │  0x00007f5f801d6ebd:   mov    %dl,0x13(%rbp,%r14,1)
          │  0x00007f5f801d6ec2:   mov    %cl,0x14(%rbp,%r14,1)        ;   {other}
   0.49%  │  0x00007f5f801d6ec7:   mov    %bl,0x15(%rbp,%r14,1)
          │  0x00007f5f801d6ecc:   mov    %r8b,0x16(%rbp,%r14,1)
   1.78%  │  0x00007f5f801d6ed1:   mov    %r9b,0x17(%rbp,%r14,1)
   0.03%  │  0x00007f5f801d6ed6:   mov    %r11b,0x18(%rbp,%r14,1)
   0.03%  │  0x00007f5f801d6edb:   mov    %r10b,0x19(%rbp,%r14,1)
          │  0x00007f5f801d6ee0:   mov    %sil,0x1a(%rbp,%r14,1)
   1.88%  │  0x00007f5f801d6ee5:   vmovd  %xmm23,%r10d
          │  0x00007f5f801d6eeb:   mov    %r10b,0x1b(%rbp,%r14,1)
   2.30%  │  0x00007f5f801d6ef0:   vmovd  %xmm24,%r10d
   0.03%  │  0x00007f5f801d6ef6:   mov    %r10b,0x1c(%rbp,%r14,1)
   0.13%  │  0x00007f5f801d6efb:   vmovd  %xmm21,%r10d
          │  0x00007f5f801d6f01:   mov    %r10b,0x1d(%rbp,%r14,1)
   0.03%  │  0x00007f5f801d6f06:   vmovd  %xmm22,%r10d
          │  0x00007f5f801d6f0c:   mov    %r10b,0x1e(%rbp,%r14,1)
          │  0x00007f5f801d6f11:   vmovd  %xmm19,%r10d
   0.03%  │  0x00007f5f801d6f17:   mov    %r10b,0x1f(%rbp,%r14,1)
   1.85%  │  0x00007f5f801d6f1c:   vmovd  %xmm20,%r10d
          │  0x00007f5f801d6f22:   mov    %r10b,0x20(%rbp,%r14,1)
   0.16%  │  0x00007f5f801d6f27:   vmovd  %xmm17,%r10d
          │  0x00007f5f801d6f2d:   mov    %r10b,0x21(%rbp,%r14,1)
   0.03%  │  0x00007f5f801d6f32:   vmovd  %xmm18,%r10d
          │  0x00007f5f801d6f38:   mov    %r10b,0x22(%rbp,%r14,1)
          │  0x00007f5f801d6f3d:   vmovd  %xmm15,%r10d
          │  0x00007f5f801d6f42:   mov    %r10b,0x23(%rbp,%r14,1)
   1.98%  │  0x00007f5f801d6f47:   vmovd  %xmm16,%r10d
          │  0x00007f5f801d6f4d:   mov    %r10b,0x24(%rbp,%r14,1)
   0.13%  │  0x00007f5f801d6f52:   vmovd  %xmm13,%r10d
          │  0x00007f5f801d6f57:   mov    %r10b,0x25(%rbp,%r14,1)
   0.10%  │  0x00007f5f801d6f5c:   vmovd  %xmm14,%r10d
          │  0x00007f5f801d6f61:   mov    %r10b,0x26(%rbp,%r14,1)
          │  0x00007f5f801d6f66:   vmovd  %xmm11,%r10d
          │  0x00007f5f801d6f6b:   mov    %r10b,0x27(%rbp,%r14,1)
   2.04%  │  0x00007f5f801d6f70:   vmovd  %xmm12,%r10d
          │  0x00007f5f801d6f75:   mov    %r10b,0x28(%rbp,%r14,1)
   0.13%  │  0x00007f5f801d6f7a:   vmovd  %xmm9,%r10d
          │  0x00007f5f801d6f7f:   mov    %r10b,0x29(%rbp,%r14,1)
   1.46%  │  0x00007f5f801d6f84:   vmovd  %xmm10,%r10d
          │  0x00007f5f801d6f89:   mov    %r10b,0x2a(%rbp,%r14,1)
   0.26%  │  0x00007f5f801d6f8e:   vmovd  %xmm7,%r10d
   0.03%  │  0x00007f5f801d6f93:   mov    %r10b,0x2b(%rbp,%r14,1)
   2.20%  │  0x00007f5f801d6f98:   vmovd  %xmm8,%r10d
   0.19%  │  0x00007f5f801d6f9d:   mov    %r10b,0x2c(%rbp,%r14,1)
   2.30%  │  0x00007f5f801d6fa2:   vmovd  %xmm5,%r10d
   0.03%  │  0x00007f5f801d6fa7:   mov    %r10b,0x2d(%rbp,%r14,1)
   1.23%  │  0x00007f5f801d6fac:   vmovd  %xmm6,%r10d
   0.06%  │  0x00007f5f801d6fb1:   mov    %r10b,0x2e(%rbp,%r14,1)
   2.14%  │  0x00007f5f801d6fb6:   vmovd  %xmm3,%r10d
          │  0x00007f5f801d6fbb:   mov    %r10b,0x2f(%rbp,%r14,1)      ;*bastore {reexecute=0 rethrow=0 return_oop=0}
          │                                                            ; - org.openjdk.bench.vm.compiler.VectorSubword::charToByte@22 (line 76)
          │                                                            ; - org.openjdk.bench.vm.compiler.jmh_generated.VectorSubword_charToByte_jmhTest::charToByte_avgt_jmhStub@15 (line 190)
   2.69%  │  0x00007f5f801d6fc0:   vmovd  %xmm4,%r11d                  ;   {other}
          │  0x00007f5f801d6fc5:   add    $0x20,%r11d                  ;*iinc {reexecute=0 rethrow=0 return_oop=0}
          │                                                            ; - org.openjdk.bench.vm.compiler.VectorSubword::charToByte@23 (line 75)
          │                                                            ; - org.openjdk.bench.vm.compiler.jmh_generated.VectorSubword_charToByte_jmhTest::charToByte_avgt_jmhStub@15 (line 190)
   0.32%  │  0x00007f5f801d6fc9:   cmp    0x34(%rsp),%r11d
          ╰  0x00007f5f801d6fce:   jl     0x00007f5f801d6d71           ;*goto {reexecute=0 rethrow=0 return_oop=0}
                                                                       ; - org.openjdk.bench.vm.compiler.VectorSubword::charToByte@26 (line 75)
                                                                       ; - org.openjdk.bench.vm.compiler.jmh_generated.VectorSubword_charToByte_jmhTest::charToByte_avgt_jmhStub@15 (line 190)
```
What is happening here?

eme64 commented:
It seems the only difference is just the level of unrolling. 8x vs 32x. But no vectorization either way.

```
public class Test {
    public static int SIZE = 1024;
    public static byte[] bytes = new byte[SIZE];
    public static char[] chars = new char[SIZE];
    
    public static void main(String[] args) {
        for (int i = 0; i < 10_000; i++) {
            test();
	}
    }

    public static void test() {
        for (int i = 0; i < SIZE; i++) {
            bytes[i] = (byte)chars[i];
	}
    }
}
```

`./java -XX:CompileCommand=compileonly,Test::test -XX:CompileCommand=printcompilation,Test::test -XX:+TraceLoopOpts -XX:-UseSuperWord Test.java`

And then it seems that the 32x unrolling leads to some interesting use of registers. I think that the issue is that first all loads are done, and we don't have enough regular registers, so we start pushing to `xmm` registers. And later move them back to regular registers. That creates a very long loop, and that is not very efficient :grimacing: 

And we somehow still don't allow vectorization of `LoadUS -> StoreB`.
@jaskarth Do you know why?

eme64 commented:
@jaskarth And of course you already filed it! I'm just playing catch-up with you :laughing: 
[JDK-8349562](https://bugs.openjdk.org/browse/JDK-8349562)
Add autovectorizer support for char casts on x86

eme64 commented:
@jaskarth Anyway, I'm super happy that you are working on patching this hole :)

jaskarth commented:
Thanks a lot for running the benchmark on your AVX512 machine! The results are very interesting, in the char cases it looks like we over-unroll the loop with SuperWord enabled even though we don't end up vectorizing the loop, fixing that could solve the slowdown. Since you mentioned the unroll amount was 32x, it might be unrolling to fill a vector (`512/sizeof(char) = 32`).

> Wait, but you seem to say that you want to support `casting to T_CHAR`. But is the issue not casting FROM char?

You are correct, I think that is my mistake. It looks like casting to char is supported because stores to both short and char become `StoreC`, but casting from char isn't supported because we have no `VectorCastC2X` node. I'll update the bug to make it more accurate.

I've also pushed a small commit to remove some extra whitespace and to make the benchmark run faster.

robcasloz commented:
> And then it seems that the 32x unrolling leads to some interesting use of registers. I think that the issue is that first all loads are done, and we don't have enough regular registers, so we start pushing to `xmm` registers. And later move them back to regular registers. That creates a very long loop, and that is not very efficient 😬

I don't know the code well but would expect `OptoRegScheduling` to mitigate this issue by producing a more register-pressure aware schedule. Do we know what is preventing that?

jaskarth commented:
@eme64 Thanks for the testing results! It looks like byte<->long conversion isn't supported with AVX1, so I've pushed a small to make the test to check for AVX2 in those cases instead.

eme64 commented:
@jaskarth Excellent! I'll run another round of testing, just to be sure :)
Please ping me again in 24h!

jaskarth commented:
@eme64 Pinging for test results :)

eme64 reviewed (approved):
Testing all passed!

Approved, thanks for the work @jaskarth !

eme64 commented:
@jaskarth I do have one more question: I have seens patterns with `LShift` and `RShift` for sign extension in the past. For example, if we cast a signed `byte` to a signed `short`, we have to do sign extension, right? Do any of these patterns appear in your benchmarks here? Do you know under which circumstances they appear? Maybe the simple conversion examples we have here are too limited to cover all cases. Not saying that should stop integration, just wondering if it is worth finding an example and filing an RFE just to keep this as a direction to explore in the future ;)

jaskarth commented:
This is a good point, while testing I experimented with patterns like this:
```java
private static short[] testSubwordVector(short[] out, int[] in) {
    for (int i = 0; i < 512; i++) {
        out[i] = (short) (((short) in[i]) + (short) in[i]);
    }

    return out;
}
```
The IR it produces looks like: `StoreC(AddI(RShiftI(LShiftI(LoadI, 16), 16)`. The same thing happens for sign extension as well. I didn't investigate too deeply, but I think the shifts prevent this pattern from vectorizing. The shifts are needed in the scalar IR since we don't have a `AddS` node, but in the future, when translating the IR to the vector graph we could convert the shift pattern into a `VectorCastX2Y` node as well.

jaskarth commented:
I searched for a bug report but didn't find one for the specific issue, so I've filed one: https://bugs.openjdk.org/browse/JDK-8357085

vnkozlov commented on src/hotspot/share/opto/superword.cpp:
```diff
@@ -2352,8 +2352,17 @@ bool SuperWord::is_velt_basic_type_compatible_use_def(Node* use, Node* def) cons
+  }
+
+  // Subword cast: Element sizes differ, but the platform supports a cast to change the def shape to the use shape.
+  if ((is_subword_type(def_bt) || is_subword_type(use_bt)) && VectorCastNode::implemented(-1, pack_size, def_bt, use_bt)) {
```
I see you use this set of conditions 2 time. Can it be separate function?  Also `-1` is strange argument for people who not familiar with code. May be add `/* comment */` to it. Or use some `#define` to have meaningful name for it.

jaskarth replied:
Thanks for the suggestion! I've moved the logic to a function to reduce code duplication and added a comment explaining the `-1`.

jaskarth commented:
/open

jaskarth commented:
I've merged from master and updated the tests to support the changes from [JDK-8350177](https://bugs.openjdk.org/browse/JDK-8350177). Now the non-truncating nodes can compile again :) I've also added the changes from the code review. Let me know what you all think!

eme64 commented on src/hotspot/share/opto/superword.cpp:
```diff
@@ -2401,8 +2401,24 @@ bool SuperWord::is_velt_basic_type_compatible_use_def(Node* use, Node* def) cons
+  return false;
+}
+
+bool SuperWord::is_supported_subword_cast(BasicType def_bt, BasicType use_bt, const uint pack_size) {
+  assert(def_bt != use_bt, "use and def types must be different");
+
+  // Opcode is only required to disambiguate half float, so we pass -1 as it can't be encountered here.
+  return (is_subword_type(def_bt) || is_subword_type(use_bt)) && VectorCastNode::implemented(-1, pack_size, def_bt, use_bt);
 }
```
Not sure if we discussed this before: should we not move this to `VectorCastNode`, rather than having it in `SuperWord`?

eme64 commented on src/hotspot/share/opto/superwordVTransformBuilder.cpp:
```diff
@@ -189,6 +189,19 @@ VTransformNode* SuperWordVTransformBuilder::get_or_make_vtnode_vector_input_at_i
+    BasicType use_bt = _vloop_analyzer.types().velt_basic_type(p0);
+
+    // If the use and def types are different, emit a cast node
+    if (use_bt != def_bt && !p0->is_Convert() && SuperWord::is_supported_subword_cast(def_bt, use_bt, pack->size())) {
```
Is `SuperWord::is_supported_subword_cast(def_bt, use_bt, pack->size())` really a true condition that you need to check here (and if false we can continue in the "else"), or should it be rather an assert?

eme64 commented on test/hotspot/jtreg/compiler/loopopts/superword/TestCompatibleUseDefTypeSize.java:
```diff
@@ -473,4 +505,201 @@ static Object[] test11(char[] a) {
+    // Narrowing
+
+    @Test
+    @IR(applyIfCPUFeature = { "avx", "true" },
+        applyIfOr = {"AlignVector", "false", "UseCompactObjectHeaders", "false"},
```
Do you think these would be supported with `asimd` as well?
If you just cannot test with it feel free to file an RFE and then I can find someone to take care of it (e.g. as a starter bug).

eme64 commented on test/hotspot/jtreg/compiler/vectorization/TestSubwordTruncation.java:
```diff
@@ -73,7 +73,7 @@ static Object[] setupCharArray() {
 
     @Test
-    @IR(counts = { IRNode.STORE_VECTOR, "=0" })
+    @IR(applyIfCPUFeature = { "avx2", "true" }, counts = { IRNode.VECTOR_CAST_I2S, IRNode.VECTOR_SIZE_ANY, ">0" })
```
Do you think we can make the vector size more precise here?

eme64 reviewed (changes requested):
I have a few more comments. This is really exciting that these cases could soon work! Thanks for working on it :blush:

eme64 commented:
@jaskarth Thanks, now the script succeeded and tests are launched :)

jaskarth commented:
@eme64 Thanks for running the tests! I think using sse4.1 here instead of avx was indeed a mistake here, and it now passes for me with UseAVX=0. I've pushed a commit fixing the bug.
